### DISCUSSION TITLE
The unique challenges of game dev regarding source control and collaboration

### DISCUSSION THREAD

**Discussion by clorl** [jj-vcs/jj#4306] [General]
Here is a summary of a discussion we had on Discord. I wanted to share my thoughts as a junior game dev on how VCS systems like Jujutsu could better address the unique challenges of game creation and the game dev industry as a whole. I've worked in small to mid-sized studios and know people who work in mid to large-scale studios. This is my humble attempt at bringing forward the unique use cases for VCS that I've been aware of and discussing 

### What makes game dev unique

Game development borrows tools from software development because the end product is indeed software, but the projects have entirely different needs regarding source control and collaboration. This is because game projects involve a wide variety of domains like art, programming, design, animation, sound, etc, which means various file types, tools, workflows, people, expertise levels, and team dynamics.

Most game projects are monolithic repos, and the game creation process revolves around maintaining a **single source of truth** at all times: the game's build. Not even the repo's head. I often say, "a feature is not finished until it's in the build."

Why is that? Because of the variety of domains involved, game development often leads to integration hell. Additionally, the software used to create games (game engines) is very complex for a myriad of reasons, one of them being that engines handle dynamic data through UIs. This means that what you see in the engine is often quite different from the final built game, which is what gets shipped to players. For example, the process of packaging assets can malfunction, causing assets to disappear or not render as they do in the engine. In multiplayer games, features might behave differently when tested locally in P2P connections, versus in a build through a P2P connection, versus in a built version with a dedicated server. Development builds, which often include cheat commands, can also behave differently from the final shipping build. The key takeaway is: **game dev is 25% making new stuff and 75% integration hell.**

Most in-development projects rely on a single source of truth (the build) and it is be the only base used for the next iterations.

One source of truth, but many different people with widely different workflows, installed toolchains, computer specs, and technical knowledge. Each person works on a specific subset of binary or text files. For example, in Perforce, there are "filters" that allow each team to pull only the relevant part of the repo, so for example programmers and designers don't have to pull the raw 3D models.

At certain stages of production or in specific types of games, there might be multiple "sources of truth," but they are all closely related to the main repository. In live games, for example, if an update is about to be released while other teams are adding content for future updates, a branch might be created from the main repo to stabilize the update. This branch becomes the "source of truth" for what is about to be delivered, while other teams continue working on future updates. Essentially, each team may need its own "source of truth" for their current work, which is challenging when working within a monorepo. I've seen cases where multiple features were developed simultaneously, but one team's changes ended up breaking another team's features.

Prototyping and testing features that might not be integrated into the final project is another area which is essential in a game's production, but where current VCS fall short. In the early stages, many prototype ideas must be tested, and these tests often cannot be cleanly separated from the main game systems. Additionally, these are not just one-person tests: they often require collaboration among multiple people and must be built and tested before a decision is made on whether to integrate them.

### Usability problems I've encountered with existing VCS

For live games, branching was extremely difficult with Perforce. Integrating changes from the main repo into a release branch was risky, and some changes involving binary files in the release branch could hardly be merged back into the main branch. This difficulty arises because game engines often make many modifications to files, even when the content hasn't changed meaningfully. This is largely due to each engine's file indexing schema. For example, Unreal Engine creates hidden "File Redirectors" if you move a file within the project, which can cause significant issues. Even renaming a file can be risky. I guess jujutsu's ability to easily make changes to parent commits could make this easier, for example by removing the change that created the file and adding a new one that creates the same file but with a different name.

This means that if you want to merge modifications, say to a buggy VFX from the release branch to the main one, you will almost certainly face conflicts. These can be resolved, but doing so may mess up the engine's file indexing, breaking references in subtle ways. I'm aware this is more of an engine-issue than VCS issue, but that's one use case game developers have to frequently face. 

File locking, like Perforce's "Checkout" feature, is useful, but making local changes to test your work is poorly managed by Perforce. If you want to make local changes to a file that a coworker has checked out, you have to click "Set as writable," which essentially makes it unwatched by Perforce. Later when you want to get the latest repo version or discard your changes, you must manually press "Reconcile Files" so that all files not being watched by Perforce are added to your changelist. Then, you need to revert all temporary files you don't want to stage and finally either overwrite, revert or integrate your local change. Iirc, if you locally marked a file as "writeable" and you pulled the latest version of the repo, Perforce would silently not pull the new version of the file and keep yours, which could simply break your local project without any clues on what happened.

### Ideas for Solutions

Paradoxically, adding **more conflicts** to VCS might improve the situation. We need to generalize and rethink VCS **merging strategies** and **conflict resolution methods**. Making "conflicts" not just about file types but also about environment or project-specific conditions could be a good idea. We may need to rethink "what is a change" and "what is a conflict".

In modern VCS, for binary conflicts, you typically have two options: "keep theirs" or "keep ours." It would be helpful to have a third option, where you can retrieve the latest file from the repo and have the VCS automatically keep your version under a different name for manual merging or to keep for prototyping purposes. You can essentially do that with jj by going to the parent commit, rename your file and retrying to merge the remote I think.

jj's support for first-class conflicts is great, but in a game dev project, for any given conflicting change, you will likely need to resolve the conflict for each team member whose local state conflicts with the remote one. This is because **the remote state of the repo rarely matches what you will have on your local machine**. To fully and rigorously resolve a conflict, you'd need to ask everyone to commit their changes, push them to the remote, pull that and resolve the conflict on one machine, then send the resolution and have everyone pull the change. An example: I renamed a texture in Unreal Engine, no conflicts, works-on-my-machine™, sent to the remote, worked on everyone else's machine, but then the texture disappeared in the build. This happened because, while pulling through the interactive interface, we manually handled things in ways that didn't cause immediate problems, but a remote headless build server didn't handle it the same way.

Redirector files often get messed up because they don't handle renaming well. In these cases, the VCS should allow you to make a manual change (e.g., renaming the file) before pulling the remote file and marking the conflict as resolved.

In video games, file references are everywhere, and are mostly not plain text and dynamic, defined by users through UIs. This means that in the event of a conflict with a binary file, even if you rename your local file and pull the remote one, references will break, leading to significant hassle. Implementing a system to "group" files together or mark files as "dependent" on other ones could be a solution. This way, when pulling or resolving conflicts, even if the files themselves haven't changed on the remote, the VCS could mark them as conflicted because a dependent file has a conflict. This could be game-changing, and I'm not sure if any VCS currently does this.

Most of the solutions I proposed above boil down to one key concept and feature that I believe no VCS has : **being able to define "what is a conflict" and "how do we resolve them"**. Let's imagine we could define "what is a conflict", and propose "conflict resolution pipelines". These "conflict definitions" could mark certain files as conflicting under certain conditions, even if we haven't locally changed the file. For example, renaming a blueprint in Unreal should be marked as a conflict, because we know it's a change that cannot meaningfully be integrated into the environment as-is. For each "conflict types", repo owners could propose "resolution pipelines". These pipelines could be shown to the user when a conflict occurs on a file, with a particular matching condition met (file type, naming schema, specific rules, etc.). The pipeline would propose one or multiple actions to resolve the conflict, defined by the repo maintainer, on top of the regular options. Users could choose which pipeline to use to resolve the conflict or resolve it themselves.

Use Case: I'm pulling a change where a binary file, e.g., an Unreal Blueprint, has been renamed. I haven't changed the file, but it's considered a conflict due to project-defined "conflict rules." My options for resolving the conflict are "keep theirs," "keep mine," "keep both," or "Resolve Rename," which runs a script that updates my project's redirectors and avoids problems.

### Conclusion

Hope this can give you an insight on what problems game dev teams face regarding source control and collaboration and to start interesting discussions. It's a bit long, so here are some key takeaways:

- Game development has unique needs due to the diversity of domains involved (art, programming, design), leading to different tools, file types, and workflows that aren't common in traditional software development.
- Monolithic repos are the norm in game development, with the game's build being the primary source of truth, rather than the latest commit in the repo.
- One source of truth but a myriad of environments and people
- We often have to work on features in parallel and try not to break other team's changes, while still being in a monorepo because branching is risky
- Game dev is 25% making new stuff and 75% integration hell
- Game engines are complex pieces of software, and the state of the codebase on any given machine is inherently non-deterministic. The remote repo's HEAD state is often different than what you have on your local machine with a clean copy.
- File locking mechanisms are cumbersome, making it difficult to locally test changes without risking unsynced files, but they maybe are essential.
- We may need to be able to think about or define per-project what is a change, what is a conflict, why two binary files, while being the same intrinsic data may need to be handled differently 
- Prototyping in game development is poorly supported by existing VCS tools, making it difficult to manage and integrate early-stage prototypes that require collaboration across teams.

### Random ideas
- There is usually two parts for game projects : The system (engine+game systems) and the content (specific gameplay implementation like level designs, abilities, but also visuals, sound etc). Usually we don't care about the structure of the project regarding content, but we do for systems. So if two people edit the same binary and commit, we really don't care if there are two of them, they might as well be duplicated along all of the dependent files. 
- So maybe there could be "alternates" changes that persist along the history? eg two people modified the player's model, instead of resolving the conflict (because we really don't care), "alternates" are created which aren't branches, but rather a new state for file changes. Alternates can easily be swapped at any point and the state of the codebase shouldn't meaningfully change regardless of which alternate is active, so selecting one locally doesn't create a commit (unless you modified the files). Then later the alternate can be manually removed or merged. This may be redundant with existing concepts but idk.
- Nix inspired source control?

**Comment by PhilipMetzger** [comment]
Here's a bag of discussions ([1][dc1], [2][dc2], [3][dc3]) from Discord,  and my related issue (#2865).

> * Nix inspired source control?

Did you mean a content addressed store? I think we can align multiple stakeholders on this feature, as it will also be nice for RBE interop with `jj run` (#1869).

[dc1]: https://discord.com/channels/968932220549103686/968932220549103689/1142153355314069544
[dc2]: https://discord.com/channels/968932220549103686/1168642196199706776
[dc3]: https://discord.com/channels/968932220549103686/968932220549103689/1225890167894839356

**Reply by clorl** [reply]
Nix inspired was just a random thought at the end. I was thinking both content adressable store and reproducable environments. I don't know the mechanisms involved in converting the commits data into the actual file structure, but you could imagine a way to construct the project files from a given state, from the commit history, just like nix config files that "reconstructs" the state of the OS everytime you change the config. This could have some benefits but that was just a random thought tbh.

**Comment by temportalflux** [comment]
For any lurking on this discussion, jj LFS support has had some progress in https://github.com/jj-vcs/jj/issues/80 and https://github.com/jj-vcs/jj/pull/9068

I think it would be great to use this discussion to also track what additional features game devs would need on top of git-LFS compatibility in order for JJ to be a competent alternative to git and perforce. I would love to return to a branch vcs with good ergonomics, but currently my team is limited to perforce for present day gamedev standards using unreal engine.

**Reply by PhilipMetzger** [reply]
> I think it would be great to use this discussion to also track what additional features game devs would need on top of git-LFS compatibility in order for JJ to be a competent alternative to git and perforce. I would love to return to a branch vcs with good ergonomics, but currently my team is limited to perforce for present day gamedev standards using unreal engine.

I mean supporting large files better than Git is like the baseline here (see #2865).

**Reply by temportalflux** [reply]
Agreed! I think thats the one issue I didnt link :laughing:. I'll be keeping an eye out as the large-files discussion and PR continues. Thanks for all your work and exploration building jj! It looks very promising.

**Comment by MultiCoreNop** [comment]
> So maybe there could be "alternates" changes that persist along the history? eg two people modified the player's model, instead of resolving the conflict (because we really don't care), "alternates" are created which aren't branches, but rather a new state for file changes. Alternates can easily be swapped at any point and the state of the codebase shouldn't meaningfully change regardless of which alternate is active, so selecting one locally doesn't create a commit (unless you modified the files). Then later the alternate can be manually removed or merged. This may be redundant with existing concepts but idk.

in Git terms this sounds like

- "exfiltrate" the affected player data to an independent repo
- have two branches there (the current player versions)
- link it in the main repo as submodule

Maybe one could web a nice UI around this?

**Reply by martinvonz** [reply]
FWIW, I really hope we will instead make it work well enough with large files in the repo that they large files don't need to be moved to a separate repo. I think we can mostly get there by "simply" providing another backend implementation. That's what we do at Google. That will work especially well when combined with a VFS so updating between commits where large files changes can be done without actually writing the large files to disk and not even reading them from the server until the user needs them.

**Reply by temportalflux** [reply]
I'm suspect at the viability of a virtual file system solution, but would be neat to see how it works. I've been in perforce with Unreal Engine for near a decade now and am very used to the "all files are local" expectation.

**Reply by PhilipMetzger** [reply]
> I'm suspect at the viability of a virtual file system solution, but would be neat to see how it works. I've been in perforce with Unreal Engine for near a decade now and am very used to the "all files are local" expectation.

It should work for Unreal and co, if it works for Meta, Google and Bungie (see https://media.gdcvault.com/GDC+2022/Speaker+Slides/Virtual+Sync+Terabytes_Moro_Brandon.pdf).

**Comment by temportalflux** [comment]
Somewhere down the line once large binary files are viable, I expect there will need to be viable engineer and non-engineer interfaces. Theres no way a team will be convinced to use jj just with command line or even a ui for engineers as the target audience. [GG](https://github.com/gulbanana/gg) is the closest thing jj has right now and it pales in comparison to UIs like [Git Fork](https://git-fork.com/), sourcetree, git kraken, etc. And even those were never particularly accessible to non-engineers.

I dont expect thats something jj itself will handle, just something the community will need in order for jj to take off in gamedev. Im definitely being tempted by my novelty drive to explore making a new UI similar to fork (my preferred git client). It'd be neat to tinker and learn jj at the same time.

Current expected toolset in ecosystems like UnrealEngine+Perforce is UnrealGameSync to "get latest" (with a sparse pull) and the in-engine integrated VCS interface for submitting assets from engine.

**Comment by clorl** [comment]
I still think that being able to define a default conflict resolution method based on some filters (file path, type, format, author, etc) would be a killer feature. Repo admins could define "conflict resolution pipelines" and tell the vcs to either use them as a default, or to prompt the user to choose.
When you think about it that's already a thing, we have the following "builtin pipelines" : 
- Manually merge text content of a file
- Use the incoming file and discard the local one
- Use the local file and discard the incoming one
- Use an external merge tool

What if for example, a team could easily define matching rules and custom default ways of resolving a conflict. This would remove a lot of manual work. Examples:

- Match rule:
    - File is a 3D model
    - Local user is one of artist_a, artist_b, artist_c
    - Incoming change has not been made by one of artist_a, artist_b, artist_c
- Resolution:
   - Discard incoming 

- Match rule:
   - File is not in the `chars/` subdirectory
   - Local user is one of artist_a, artist_b, artist_c
- Resolution:
   - Discard local changes
